### DISCUSSION TITLE
How to CDK for Service Catalog Products

### DISCUSSION THREAD

**Discussion by TannerIsBender** [aws/aws-cdk#15350] [Q&A] [answered]
Hello,

I'm learning how to use the CDK with Java and have come up with a project to help me learn. My plan is to use the CDK to create CI/CD for updating products within a portfolio using the AWS Service Catalog.

My first step in this project is to create a portfolio and add products via the CDK/CloudFormation. I was able to create the portfolio so I started working on adding a product, but this is where I ran into trouble...

cdk --version -> 1.110.0 (build 32f00eb)

I have started creating a product, but am running into a problem with creating the provisioningArtifactParameters

```
CfnCloudFormationProduct product = new CfnCloudFormationProduct(portfolio, "someId", CfnCloudFormationProductProps.builder()
                .name("Some name")
                .description("Some description")
                .owner("i am groot")
                .provisioningArtifactParameters() // need help here
                .build());
```

I found [documentation for the provisioningArtifactParameters](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-servicecatalog.CfnCloudFormationProduct.ProvisioningArtifactPropertiesProperty.html) but I don't see any examples for how to do it Java.

Hopefully that is enough information for what I'm trying to accomplish... Now I'll show you what I tried and failed at.

### Trying to create a product:

```
        List<ProvisioningArtifact> provisioningArtifact = new ArrayList<>();
        Object info = "{ LoadTemplateFromURL: asset.s3Url }";
        provisioningArtifact.add(new ProvisioningArtifact(info));

        CfnCloudFormationProduct product = new CfnCloudFormationProduct(portfolio, "someId", CfnCloudFormationProductProps.builder()
                .name("Some name")
                .description("Some description")
                .owner("i am groot")
                .provisioningArtifactParameters(provisioningArtifact)
                .build());
```

### Implementing the interface
```
import software.amazon.awscdk.services.servicecatalog.CfnCloudFormationProduct;

public class ProvisioningArtifact implements CfnCloudFormationProduct.ProvisioningArtifactPropertiesProperty {

    public String description;
    public String version;
    public Object info;

    public ProvisioningArtifact(Object info) {
        this.info = info;
        this.description = "testDescription";
        this.version = "testVersion";
    }

    public ProvisioningArtifact(String description, String version, Object info) {
        this.description = description;
        this.version = version;
        this.info = info;
    }

    @Override
    public Object getInfo() {
        return info;
    }

    @Override
    public String getDescription() {
        return description;
    }

    @Override
    public String getName() {
        return "MyNameHere";
    }

}
```

I appreciate any help or comments to help me resolve this issue! Thank you

Additional information: https://stackoverflow.com/questions/68169398/aws-cdk-java-cfncloudformationproduct-provisioningartifactparameters

**Comment by arcrank** [comment, marked as answer]
Hi @TannerIsBender.  We actually will have an L2 version of the cloudformation product coming out with the next release that should be easier to use, see commit [here](https://github.com/aws/aws-cdk/commit/fe3e0f2058ce48413eca88b247cddf0532374a74).  We have simplified the provisioning artifacts somewhat into _product versions_ which will make things clearer.  I will dig into this in meantime though.

**Reply by TannerIsBender** [reply]
Thank you for the reply! That release is exactly what I'm looking for :) 

In the meantime, please let me know if you find a temporary solution.
